File: facilitylocation/facility_location.py
########## facility_location.py ##########

import localsolver
import sys

if len(sys.argv) < 2:
    print("Usage: python facility_location.py inputFile [outputFile] [timeLimit]")
    sys.exit(1)


def read_integers(filename):
    with open(filename) as f:
        return [int(elem) for elem in f.read().split()]


with localsolver.LocalSolver() as ls:

    #
    # Reads instance data
    #

    file_it = iter(read_integers(sys.argv[1]))
    # Number of locations
    N = next(file_it)
    # Number of edges between locations
    E = next(file_it)
    # Size of the subset S of facilities
    p = next(file_it)

    # w: Weight matrix of the shortest path between locations
    # wmax: Maximum distance between two locations
    wmax = 0
    w = [None]*N
    for i in range(N):
        w[i] = [None]*N
        for j in range(N):
            w[i][j] = next(file_it)
            if w[i][j] > wmax:
                wmax = w[i][j]

    #
    # Declares the optimization model
    #
    m = ls.model

    # One variable for each location : 1 if facility, 0 otherwise
    x = [m.bool() for i in range(N)]

    # No more than p locations are selected to be facilities
    opened_locations = m.sum(x[i] for i in range(N))
    m.constraint(opened_locations <= p)

    # Costs between location i and j is w[i][j] if j is a facility or 2*wmax if not
    costs = [None]*N
    for i in range(N):
        costs[i] = [None]*N
        for j in range(N):
            costs[i][j] = m.iif(x[j], w[i][j], 2*wmax)

    # Cost between location i and the closest facility
    cost = [None]*N
    for i in range(N):
        cost[i] = m.min(costs[i][j] for j in range(N))

    # Minimize the total cost
    total_cost = m.sum(cost[i] for i in range(N))
    m.minimize(total_cost)

    m.close()

    #
    # Parameterizes the solver
    #
    if len(sys.argv) >= 4: ls.param.time_limit = int(sys.argv[3])
    else: ls.param.time_limit = 20

    ls.solve()

    #
    # Writes the solution in a file following the following format:
    # - value of the objective
    # - indices of the facilities (between 0 and N-1) */
    #
    if len(sys.argv) >= 3:
        with open(sys.argv[2], 'w') as f:
            f.write("%d\n" % total_cost.value)
            for i in range(N):
                if x[i].value == 1:
                    f.write("%d " % i)
            f.write("\n")

File: facilitylocation/Facilitylocation.cs
/********** Facilitylocation.cs **********/

using System;
using System.IO;
using System.Collections.Generic;
using localsolver;

public class Facilitylocation : IDisposable
{
    // Number of locations
    int N;

    // Number of edges between locations
    int E;

    // Size of the subset S of facilities
    int p;

    // Weight matrix of the shortest path beween locations
    int[][] w;

    // Maximum distance between two locations
    int wmax;

    // LocalSolver
    LocalSolver localsolver;

    // Decision variables
    LSExpression[] x;

    // Objective
    LSExpression totalCost;

    // List of facilities
    List<int> solution;

    public Facilitylocation()
    {
        localsolver = new LocalSolver();
    }

    // Reads instance data
    public void ReadInstance(string fileName)
    {
        using (StreamReader input = new StreamReader(fileName))
        {
            var tokens = input.ReadLine().Split(' ');
            N = int.Parse(tokens[0]);
            E = int.Parse(tokens[1]);
            p = int.Parse(tokens[2]);

            w = new int[N][];

            wmax = 0;

            for (int i = 0; i < N; i++)
            {
                tokens = input.ReadLine().Split(' ');
                w[i] = new int[N];
                for (int j = 0; j < N; j++)
                {
                    w[i][j] = int.Parse(tokens[j]);
                    if (w[i][j] > wmax)
                        wmax = w[i][j];
                }
            }
        }
    }

    public void Dispose()
    {
        if (localsolver != null)
            localsolver.Dispose();
    }

    // Declares the optimization model
    public void Solve(int limit)
    {
        localsolver = new LocalSolver();
        LSModel model = localsolver.GetModel();
        x = new LSExpression[N];

        // One variable for each location : 1 if facility, 0 otherwise
        for (int i = 0; i < N; i++)
            x[i] = model.Bool();

        // No more than p locations are selected to be facilities
        LSExpression openedLocations = model.Sum(x);
        model.Constraint(openedLocations <= p);

        // Costs between location i and j is w[i][j] if j is a facility or 2*wmax if not
        LSExpression[][] costs = new LSExpression[N][];
        for (int i = 0; i < N; i++)
        {
            costs[i] = new LSExpression[N]; 
            for (int j = 0; j < N; j++)
            {
                costs[i][j] = model.If(x[j], w[i][j], 2 * wmax);
            }
        }

        // Cost between location i and the closest facility
        LSExpression[] cost = new LSExpression[N];
        for (int i = 0; i < N; i++)
            cost[i] = model.Min(costs[i]);

        // Minimize the total cost.
        totalCost = model.Sum(cost);
        model.Minimize(totalCost);

        model.Close();

        // Parameterizes the solver
        localsolver.GetParam().SetTimeLimit(limit);
        localsolver.Solve();

        solution = new List<int>();
        for (int i = 0; i < N; i++)
            if (x[i].GetValue() == 1)
                solution.Add(i);
    }

    // Writes the solution in a file following the following format: 
    // - value of the objective
    // - indices of the facilities (between 0 and N-1)
    public void WriteSolution(string fileName)
    {
        using (StreamWriter output = new StreamWriter(fileName))
        {
            output.WriteLine(totalCost.GetValue());
            for (int i = 0; i < solution.Count; ++i)
                output.Write(solution[i] + " ");
            output.WriteLine();
        }
    }

    public static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: Facilitylocation inputFile [outputFile] [timeLimit]");
            Environment.Exit(1);
        }
        string instanceFile = args[0];
        string outputFile = args.Length > 1 ? args[1] : null;
        string strTimeLimit = args.Length > 2 ? args[2] : "20";

        using (Facilitylocation model = new Facilitylocation())
        {
            model.ReadInstance(instanceFile);
            model.Solve(int.Parse(strTimeLimit));
            if (outputFile != null)
                model.WriteSolution(outputFile);
        }
    }
}
